### ISSUE TITLE
Delay/freeze when switching focus & workspace

### ISSUE THREAD

**Issue by fmasa** [nikitabobko/AeroSpace#131]
Hi!
Thanks for all the new features.

After updating to 0.8.0, I had AeroSpace disabled before I fixed the config file and I noticed that when AerosSpace is enabled there is a noticeable delay (I have no measurements, but it feels like hundreds of milliseconds at least) when either moving focus or changing workspaces. This also happens when I focus windows via `cmd+tab`, doing this without AeroSpace is instantaneous.

I know it's still beta, but I wanted to ask:

- Do other people experience this delay or am I missing some configuration
- Is this a performance issue in AeroSpace or some inherent limitation of Accessibility API?
- Is there any way to collect logs/diagnostics to find out more?

I'm on M1 Pro + Sonoma.
 
---

I'm not trying to bash AeroSpace - the feature set is great and I use it as a daily driver, but the MacOS felt laggy for a while and now I found out it's only when using AeroSpace and I want to help to make it better.

---

Note by @nikitabobko

Related: #242

**Comment by nikitabobko** [comment]
**(1)** How many windows do you have opened? (you can use `aerospace list-windows --all | wc -l`) 
**(2)** Does the AeroSpace still behave poorly if you restart the app?

I saw on my colleague's machine that AeroSpace may behave poorly if there are unreasonably many windows opened (maybe 100 or even more, we didn't count)

I tried to open 20 windows and it performs well, I don't see any delay compared to when AeroSpace is not running

> Do other people experience this delay or am I missing some configuration

I can't see any delay on my machine (M1 Pro + Sonoma). I normally don't have more than 10 windows opened

> or some inherent limitation of Accessibility API?

I heard several times that Accessibility API is slow. Though, I can't confirm it yet. The only real slowness of Accessibility API I noticed is when some apps block their UI thread _in some special way_ (cause I couldn't reproduce the slowness when I wrote a small app that displays a simple window and blocks the UI thread).

For example, Spotify and Firefox sometimes block the UI thread _in a special way_ for several seconds when I try to quit (`cmd+q`) either of them, Accessibility API is blocked by these apps until they finally quit. I've not yet debugged this problem. 

**(3)** Some of the apps that you are running may have similar problems, so you can try to quit apps one by one to see when AeroSpace delay goes away

**Comment by Kamyil** [comment]
Hi! I'm experiencing this issue on my Macbook Pro 14 with M1 Pro on macOS Sonoma, but a little bit differently.
When I'm not switching workspaces for a while, then first workspace switch has a little bit of delay
_(it's hard to measure that, I'm front-end dev dealing with transitions/animations day-by-day, so my eyes are a little bit trained, but not perfect ofc)_

but when I switch and then immediately switch again and again (and so on) it takes noticeably less time to switch. 
So it behaves like it does some kind of cool start on first switch and then it runs faster after the first switch, creating illusion that it has some delay on first switch

If there was some logging tool that would allow to measure the time to switch workspace, then I could attach such log file to this comment. It's actually hard to record on video without some keys recording (but I can try ofc)

**Comment by F-TD5X** [comment]
Hi! I'm experiencing this issue too on  M1 Pro MBP.

And the situation only happened when I starting debug an application with idea. 

Seems this is what @nikitabobko said but a stable reproduce way. (at least for me)
```
IntelliJ IDEA sometimes blocks the UI thread in a way that blocks accessibility API (I don't know the reproducer yet)
```
Are there any way I can offer more infos?

**Comment by fmasa** [comment]
I have IDEA opened 90+% of time, so that might be a culprit 🤔

**Comment by fmasa** [comment]
I encountered it a few times again and killing IntelliJ IDEA fixes the issue. It seems that AeroSpace commands are queued because of all the workspace switching I tried to do when it hung are quickly applied after the IDEA is killed.

**Comment by nikitabobko** [comment]
The potential fix is to send accessibility requests to different applications from different threads. This way, if one application hangs, it won't affect others. But I don't know whether Apple accessibility API is thread-safe (My intuition says that it's probably not)

Also, I'm curious whether yabai/Amethyst have this problem (obviously, if they have this problem, it's not as severe for them, because they use native macOS spaces, but AeroSpace needs to constantly move windows to the bottom right corner and back)

**Comment by cannonrush** [comment]
Somewhat related, since it also introduces a noticeable delay: is it possible to execute all commands (such as changing focus or switching workspaces) on key press instead of key release?

**Comment by nick-potts** [comment]
I've been getting it quite a bit lately.My cpu is often maxed out at 100%, often by a bash process running in my IDEA terminal, and my window management then has a 4-10s delay. 

It's easily replicable with `stress -c 16` (or whatever cores your Mac has) installed from brew. Interestingly, from a regular terminal this same behaviour doesn't seem to happen.

> Also, I'm curious whether yabai/Amethyst have this problem (obviously, if they have this problem, it's not as severe for them, because they use native macOS spaces, but AeroSpace needs to constantly move windows to the bottom right corner back and forth)
I believe from memory they did, but "spaces" was always fine under heavy cpu load.

**Comment by gohma231** [comment]
I get this a lot as well. I frequently have high CPU load on my machine and aerospace is very delayed in switching.

**Comment by diogox** [comment]
Same here!

The 4-10s delays are a real struggle :(

Never made the connection to IntelliJ, but can confirm: I daily Goland (it's IntelliJ under the hood) and have been noticing huge slowdowns - potentially as I open more instances(?) - but mostly when I have > 1 monitors connected.

**Comment by nick-potts** [comment]
Just have to run `stress -c 16` in the integrated terminal in an IntelliJ app and you'll be in a world of pain.

**Comment by gempir** [comment]
Also constant IntelliJ user. Sometimes I notice like 10-15s delay and for some reason quitting Aerospace and reopening does help a bit.

**Comment by Hritik14** [comment]
Lag with intelij running is a lot that it hinders usability. when I switch to a different workspace, I see my background for a few seconds then all the windows render on the workspace. This is also visible when IDEA is not running but it is very quick.

**Comment by kahnclusions** [comment]
I've also noticed this issue, and I'm not an Intellij user. Sometimes I can narrow down a specific app (for example I noticed Swift Mail causes a lot of lag), but sometimes I just have Firefox, Wezterm, and Discord open and still get a 5-10 second lag. Restarting Aerospace helps for a while until it comes back.

It's a pity because I really love Aerospace and have been daily driving it for the past 2-3 months!

**Comment by chinedufn** [comment]
I am able to reliably reproduce the problem.

Background: I'm a new user. Been using aerospace for about a month and change.

---

After dealing with the delaying or freezing for my first two or so weeks of using `aerospace`, I just I went a couple of weeks without any delaying/freezing.

I went as far as checking to see if there was a new aerospace release that got auto-updated or something, since it had gone from slow to fast out of nowhere.

I didn't see a new release. That confused me.

---

Anywho, after that couple of weeks without any performance issues I ran my codebase's test suite, which is CPU intensive.

Aerospace immediately became slow, and remained slow even after my CPU load went back to 1-2%.

Once I restarted it, it was fast again.

Looking back, for the two or so weeks where it was fast I had:

1. Restarted aerospace because I was restarting my computer

2. Not done anything CPU intensive for a couple of weeks

3. As soon as I did something CPU intensive, it became slow again. Restarting aerospace made workspace switching fast again.

---

The above suggests that this issue is not directly related to what applications you are running.

**The slowness begins as soon as you do something CPU intensive, regardless of what application caused it.**

People are reporting that they notice the problem when running intellij.

I suspect that this is because intellij is something that people use to do CPU intensive things.

---

Given that the problem immediately goes away once you restart aerospace, I'm inclined to believe that aerospace is capable of either fixing or working around whatever the problem is.

For instance, in the worst case aerospace could detect unusual slowness and do whatever it does when it restarts. (I am not suggesting that this is a good idea, just pointing one a brute force solution).

In the best case, we identify what aspects of aerospace could at all be related to this issue, narrow down the root cause, and fix it.

---

In summary:

- the above describes how to reproduce this issue

- the issue seems solvable within aerospace, be it using a brute force solution or a more elegant one

---

Performance issues have been on an intel mac.
```
2.4 GHz 8-Core Intel Core i9
2019 MacBook Pro
```

**Comment by gempir** [comment]
Today I found it goes so far that Activity monitor says Aerospace is not responding and if you hover the menubar item of Aerospace you get the beachball. And you could do zero actions with Aerospace.

This happend because IntelliJ somehow also got stuck and had to be force-quit. I even restarted Aerospace (while IntelliJ was still in its broken state) and Aerospace was barely responsive only resizing some windows, switching workplaces didn't work at all.

Only fully force quitting IntelliJ brought life back to Aerospace.

**Comment by nikitabobko** [comment]
No need to report about applications that block macOS AX API (such as IntelliJ, or Emacs `sleep 5`). It's known problem and I can reproduce it. There are several directions to investigate

1. Make use of `AXUIElementSetMessagingTimeout`, and implement some kind of backoff strategy for applications that hang
2. Dedicate a thread for each application and issue AX requests from them (IDK if it's a thread-safe)

Related yabai discussions: https://github.com/koekeishiya/yabai/issues/2377 https://github.com/koekeishiya/yabai/issues/599 https://github.com/koekeishiya/yabai/issues/600

Until I have time to dig into this problem, everyone else is welcome to do so

**Comment by GuiltiTer** [comment]
Same issue here!
I have an M2 Pro machine. Today, with the Telegram desktop app, Kitty, and IINA windows, I opened. Killing all applications one by one and restarting the Aerospace didn't fix the workspace switch lagging (sometimes not responding in a few minutes). Killing the Finder did the job, though.

**Comment by nagy135** [comment]
i also have hard time pinpointing what the issue is ...yesterday i had like 3sec freezes switching desktops and closed everything but browser and alacritty and it fixed it. For me its one of (or all combined) BambuStudio, Blender, Spotify desktop.

**Comment by emielvangoor** [comment]
I am experiencing the same issues. Not using Intellij but i suspect its due to the Arc browser. Aerospace is snappy in the beginning but quickly gets very, _very_ slow, almost unusable. Restarting Aerospace does the trick.

M1 Pro on Sequoia 15.0

**Comment by kahnclusions** [comment]
I think we know what the root of the issue is, based on [this comment](https://github.com/nikitabobko/AeroSpace/issues/131#issuecomment-2302135213), the open question at this point is what can be done to solve it.

**Comment by emielvangoor** [comment]
> I am experiencing the same issues. Not using Intellij but i suspect its due to the Arc browser. Aerospace is snappy in the beginning but quickly gets very, _very_ slow, almost unusable. Restarting Aerospace does the trick.
> 
> M1 Pro on Sequoia 15.0
> 
> Edit; It seems when I disable JankyBorders its working better. Time will tell.

After full working day I can say this did the trick, disabling JankyBorders fixed all issues for me..

**Comment by michaelessiet** [comment]
I experience this issue even without jankyborders. I think the threading thing might be the issue because once I restart my computer aerospace works fine at least until I get a couple apps running for an extended period of time (a few days) then I experience the lag.

**Comment by nagy135** [comment]
i too see the same, but whenever it starts degrading, all i need to do is close all the apps (excluding terminal and browser) so probably some apps are causing this to be worse but it always fixes it for me. Then its almost instant for days, then there seems to be a point where it suddenly starts degrading again and turns from fraction of second to few seconds. Then closing apps solves it again.

**Comment by nikitabobko** [comment]
In 0.15.0, I added a low effort commit to set `AXUIElementSetMessagingTimeout` to 1 second.

It won't resolve the problem, but should avoid complete freezes caused by 3rd party apps

**Comment by KasselMarc** [comment]
I also experience aerospace getting slower over time, not as freezing completely but definitely some delays 1-2s delays having around 10 windows at 10 different work spaces open. I don't use JankyBorders.

**Comment by ronaldsuwandi** [comment]
> In 0.15.0, I added a low effort commit to set `AXUIElementSetMessagingTimeout` to 1 second.
> 
> It won't resolve the problem, but should avoid complete freezes caused by 3rd party apps

Used the latest 0.15.2 but still face the similar problem. It's particularly prevalent when having high CPU usage (IntelliJ compiling). Switching between workspaces can take seconds. Really hope this get resolved soon 🙏

**Comment by raisjn** [comment]
I was able to reduce (read: eliminate) the multi-second stalls by pushing work to a separate thread and can now push to 100% CPU without getting any slow down in Aerospace. I test drove it for the past day and it's working well so far. Will see if any problems surface over the next couple days but it makes me optimistic that this problem is capable of being solved.

**Comment by saivan** [comment]
I seem to be getting the exact same problem from chromium based apps. Happens when I push CPU usage on both Arc and VSCode. Delays can be 20 seconds. I'm on an M1 Max most of the time

**Comment by srgsanky** [comment]
> In 0.15.0, I added a low effort commit to set AXUIElementSetMessagingTimeout to 1 second.

> It won't resolve the problem, but should avoid complete freezes caused by 3rd party apps

@nikitabobko How can I use the latest version with this fix?

```
> brew upgrade aerospace
Warning: Not upgrading aerospace, the latest version is already installed

> aerospace --version
aerospace CLI client version: 0.14.2-Beta 0cb8dbdfc5ee73b8cbc200f175f467ebead55201
AeroSpace.app server version: 0.14.2-Beta 0cb8dbdfc5ee73b8cbc200f175f467ebead55201
```

This is using the commit from Aug 14 https://github.com/nikitabobko/AeroSpace/commit/0cb8dbdfc5ee73b8cbc200f175f467ebead55201

**Comment by jack4git** [comment]
I missed this ticket and created #604 -- lots of details there on my experience and config, same issue I'm sure.

**Comment by clarsen** [comment]
i'm seeing that the detectNewWindowsAndAttachThemToWorkspaces is slow, but on further delving into it, the iteration over apps MacApp.get() construction/population is slow for just one of the long running apps.  When I closed that app, everything sped up again.

**Comment by nick-potts** [comment]
> I was able to reduce (read: eliminate) the multi-second stalls by pushing work to a separate thread and can now push to 100% CPU without getting any slow down in Aerospace. I test drove it for the past day and it's working well so far. Will see if any problems surface over the next couple days but it makes me optimistic that this problem is capable of being solved.

Do you mind pushing your changes to your fork - I'd love to test drive as well.

**Comment by raisjn** [comment]
@nick-potts: i've been moderately busy, but i started a discussion with a list of approaches I've tried: https://github.com/nikitabobko/AeroSpace/discussions/655. I will see when I can find time to polish the code a little and put it in a fork - I have several modifications that are separate from this and would need to extract just this change.

**Comment by liamaharon** [comment]
@raisjn appreciate you working on this! I will gladly offer 500 USD (paid in ETH) if you merge a fix here that keeps aerospace responsive while my CPU is saturated compiling code.

**Comment by raisjn** [comment]
@nick-potts: I put a link to a branch in #655, please try it out and remember to enable it with `new-window-detection-timeout = 100` config option in aerospace.toml. There are several potential bugs - see the listed ones in #655 and if you notice others, please add them to the discussion. 

@liamaharon: thank you for the offer, i appreciate the thought but can't accept - feel free to sponsor aerospace instead.

**Comment by zapoutix** [comment]
Hello @raisjn 
I just compile the branch "dev/okay/2024_11_02_less_cpu_lockups" and add new-window-detection-timeout = 100 in config file.
When Godot is open is still laggy when switching workspace.

**Comment by raisjn** [comment]
@zapoutix, @liamaharon: thanks for reporting back and sorry there is still some slowness. there's two config options that can be played with: `new-window-detection-debounce` and `new-window-detection-timeout`. When you raise debounce, it will cause less work to happen, when you lower timeout, it will wait less time for the detect new window logic to finish.

You can try `new-window-detection-debounce 100` (or 200), or you can try lowering `new-window-detection-timeout` to 50. I don't have a repro on slowness, so its difficult for me to debug - if you are able to get any further insight into which piece is running slow (adding print statements with timing within `refreshSession`?), that would give me some ideas!

**Comment by zapoutix** [comment]
@raisjn i will have a try,
with the "dev/okay/2024_11_02_less_cpu_lockups" version i got a crash, never happened with the 15.2,

`##### AeroSpace Runtime Error #####

Please report to:
    https://github.com/nikitabobko/AeroSpace/issues/new
    Please describe what you did to trigger this error

Message: Windows always have parent
Version: 0.0.0-SNAPSHOT
Git hash: bd23ebece513751c0ca00a18a345bddf1f75bb41
Coordinate: /Users/mattal/Downloads/AeroSpace/Sources/AppBundle/tree/Window.swift:7:72 parent
recursionDetectorDuringFailure: false
cli: false

Stacktrace:
0   AeroSpace                           0x00000001005931d0 $s6Common6errorT_4file4line6column8functionxSS_SSS2iSStlF + 744
1   AeroSpace                           0x0000000100436c28 $s9AppBundle19tryOnWindowDetected_7startupyAA0E0C_SbtF + 160
2   AeroSpace                           0x0000000100436d90 $s9AppBundle9MacWindowC3get3app02axD07startupACSgAA0cA0C_So14AXUIElementRefaSbtFZyyScMYccfU_ + 16
3   AeroSpace                           0x000000010041eb5c $sIegh_IeyBh_TR + 28
4   libdispatch.dylib                   0x000000019211a8f8 _dispatch_call_block_and_release + 32
5   libdispatch.dylib                   0x000000019211c658 _dispatch_client_callout + 20
6   libdispatch.dylib                   0x000000019212af68 _dispatch_main_queue_drain + 980
7   libdispatch.dylib                   0x000000019212ab84 _dispatch_main_queue_callback_4CF + 44
8   CoreFoundation                      0x00000001923f4e60 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 16
9   CoreFoundation                      0x00000001923b4a4c __CFRunLoopRun + 1996
10  CoreFoundation                      0x00000001923b3bc4 CFRunLoopRunSpecific + 588
11  HIToolbox                           0x000000019d823f64 RunCurrentEventLoopInMode + 292
12  HIToolbox                           0x000000019d829d54 ReceiveNextEventCommon + 636
13  HIToolbox                           0x000000019d829eb8 _BlockUntilNextEventMatchingListInModeWithFilter + 76
14  AppKit                              0x0000000195edfa08 _DPSNextEvent + 660
15  AppKit                              0x000000019681fe0c -[NSApplication(NSEventRouting) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 688
16  AppKit                              0x0000000195ed2ae0 -[NSApplication run] + 480
17  AppKit                              0x0000000195ea9364 NSApplicationMain + 888
18  SwiftUI                             0x00000001c0558340 $s7SwiftUI6runAppys5NeverOSo21NSApplicationDelegate_So11NSResponderCXcFTf4e_nAA07TestingdG0C_Tg5Tm + 160
19  SwiftUI                             0x00000001c09c8ee4 $s7SwiftUI6runAppys5NeverOxAA0D0RzlF + 84
20  SwiftUI                             0x00000001c0d41f24 $s7SwiftUI3AppPAAE4mainyyFZ + 224
21  AeroSpace                           0x00000001003ba474 main + 36
22  dyld                                0x0000000191f4c274 start + 2840
`

**Comment by raisjn** [comment]
@zapoutix: thanks for reporting! i'm familiar with this crash :'( - see the update in #655 - can you move your comment + future comments on behavior of the patch to #655? (same for anyone else discussing the patch)

**Comment by nikitabobko** [comment]
Today I received a very generous $500 donation via GitHub Sponsorship from @liamaharon. I appreciate the support, thank you! It's the highest individual sponsorship I received, so it makes it clear how important this issue for you.

As for the issue, nothing significantly changed since my last update https://github.com/nikitabobko/AeroSpace/issues/131#issuecomment-2302135213 

I understand the importance of this issue. Discussions, hacks people put in AeroSpace sources and the $500 donation make it clear. But I won't be giving any promises. The issue is fixed when it's fixed if it's fixed at all. Thanks everyone for your understanding

---

The core architecture of AeroSpace is built around what I call `refreshSession`. It's an idempotent procedure that is run on every relevant event from macOS or the user itself (e.g. keybind). On every `refreshSession`, AeroSpace checks for newly appeared windows, checks that windows are still in the right position, and pushes them ~~down the macOS throat~~ in the right position.

This approach makes sure that we don't rely on reliability of macOS events being delivered to AeroSpace at all or being delivered in the right order (yes, macOS can just drop the event and not deliver it to apps, or if it does deliver it, it will very likely be in inconsistent order). Until macOS sends events that at least approximately look like something reasonable, AeroSpace can provide a reliable experience. I was pissed off by how often some other window managers loose track of windows, so I went with this hard approach, and very happy about it. Hopefully, you notice that AeroSpace *never* looses track of windows.

\<rant>I took a shortcut and didn't implement the detection of closed windows in the same idempotent manner, because when the screen is locked, closed windows become indistinguishable from non-closed windows. Unironically, Apple decided to break specifically `kAXUIElementDestroyedNotification` event in Sequoia #445, ughh\</rant>

The flip side of the "idempotent"/"recheck the whole world every time" approach is performance. But I think it can be optimized, it just something that needs to be explicitly worked on, and worked very accurately without harming the correctness. I never spent time to properly optimize AeroSpace and macOS communication, so far I've been concerned only about the correctness and reproducibility (but still managed to get bitten by Sequoia #445).

Thanks to @raisjn for doing his performance experiments. The debouncing approach makes sense. The `refreshSession` may be called too often when a lot of things happen on the screen (e.g. when you switch the workspace), and given the idempotent nature of `refreshSession`, debouncing the whole session becomes trivial and may increase the app responsiveness. I don't agree with the specifics of implementation, but the general idea makes sense, thanks for your experiments. I will consider "the debouncer" approach when start working on performance.

Another known problem that I realized is that AeroSpace cares too much about correctness for *invisible* windows, and makes too much accessbility API requests because of that #545

As correctly pointed by @raisjn, please discuss the details of @raisjn fix in #655

Please, restrain from casually commenting in this issue, since it becomes unnecessarily long. If you want to say something, please say something that we don't know already.

**Comment by camiloaromero23** [comment]
I didn't read the issue completely 'cause it is a bit long. But I noticed that when I closed some opened windows (I have a MBP M1 Pro w 16GB RAM), it continued working flawlessly. Maybe it is an issue with memory usage (swap). Hope this gives valuable information for this to be fixed. Thanks for the nice work with aerospace 🚀

**Comment by stevo-knievo** [comment]
> In fact, coming here, I just need the ability to switch desktops instantly, I don't really care about auto-layout, when I'm working I manually set the window up, it takes no more than a few seconds and I don't move it around anymore.
> If there was a simplified version that only offered the ability to switch desktops, no longer cared about layout, and performance was no longer an issue, then I'd be happy!
> 

You don't need Aerospace to do that. In macOS you can set up keyboard shortcuts to switching desktops.

**Comment by nagy135** [comment]
> > In fact, coming here, I just need the ability to switch desktops instantly, I don't really care about auto-layout, when I'm working I manually set the window up, it takes no more than a few seconds and I don't move it around anymore.
> > If there was a simplified version that only offered the ability to switch desktops, no longer cared about layout, and performance was no longer an issue, then I'd be happy!
> 
> You don't need Aerospace to do that. In macOS you can set up keyboard shortcuts to switching desktops.

not rly, it has animations and thats the reason we meet here in the first place. If there was bind to instantly switch to another macos WS then we would not need anything else.

Recently i switched to raycast with binds for specific programs and its honestly what i wanted always. If you rarely have more than one window per WS, raycast will be all you need ....and its instant (and even opens app if its not running 🤤)

**Comment by nikitabobko** [comment]
Update on the issue.

**1.** I've tried to call macOS blocking AX API from different threads for different applications a while ago, it worked and didn't crash, so my hope is that it can work this way. Unfortunately, Apple documentation doesn't answer on that question

**2.** I'm in the process of migrating the core of AeroSpace to use thread-per-application approach. Meaning that hanging slow apps won't block the rest of the system. It's the core of the app, so the migration will take time.

Coincidentally, the latest big feature in Swift community is concurrency-strict-mode which promises to eliminate data-races. I've played with this feature, and I'm still learning its corner-cases, the feature comes in handy for our thread-per-application purposes. AeroSpace has already migrated to Swift 6 in the `main` branch.

**3.** Since I'm rewriting some of the core parts of macOS communication, I am in the process of making *the refresh session* cancellable (I explained what the refresh session is earlier in this thread). Given the idempotent nature of refresh session, when a new events is received, it can cancel the currently in progress session, which will prevent AeroSpace from doing unnecessary work. It is effectively `@raisjn`'s approach but without losing "the initial responsiveness"

**4.** I also consider to split big serial "refresh session" architecture to different types of parallel sessions/streams/event-handlers approach. The idea is that the events that come directly from users (CLI requests and key codes handling) should be processed as quick as possible, so we should do only the essential job in a *hot event-handler* (only handle the event, and place windows in the correct position. For query commands, we don't even need to place windows in the correct position). Dead windows detection and new windows detection (which are the most time consuming) could be excluded from this hot event-hanlder into a background job

---

I will give an update on the migration to thread-per-application approach whether it succeeds or fails.

This issue stays locked to avoid non-constructive comments

**Comment by nikitabobko** [comment]
I have pushed the initial version of mentioned optimizations to `main` branch. The main commits are https://github.com/nikitabobko/AeroSpace/commit/27689c6bfb153c320e03c9becfb54da0a4df4077 & https://github.com/nikitabobko/AeroSpace/commit/05a34f79009ae2ccbc88f8d6c8ad7980b1a7aa48

I've checked the following scenarios:

**1.** Attach debugger to Java GUI application. Previously, it'd entirely block AeroSpace, now it's fine. AeroSpace will even detect such window as dead (because it doesn't respond to any AX requests) and exclude it entirely

**2.** Python blocking `pygame`
```python
import pygame
import time
pygame.display.init()
time.sleep(5000)
```
The same result as in 1.

**3.** CPU stress test. Build kotlin-native from sources. Previously, it'd make AeroSpace laggy. Now AeroSpace is as responsive as during idle CPU

**4.** Godot. This one is the worst. You can still notice a small maybe half second workspace switch lag. But it's much more usable, and AeroSpace doesn't freeze. I'd like to take a look specifically at Godot later – tracking issue https://github.com/nikitabobko/AeroSpace/issues/218

---

I'd like to test AeroSpace for a week, process PRs, and hopefully will be able to do the release next weekend. Those who eager to try the fix earlier can do that via `./install-from-sources.sh` in the repo root (it will install AeroSpace as `aerospace-dev` homebrew cask)

If you have any questions in regard to `./install-from-sources.sh` please use GH Discussions

I've opened the issue for commenting. If you try the fix from `main` branch, feel free to comment whether it resolves or doesn't resolve your scenarios

**Comment by lazywei** [comment]
I uploaded the build artifact for macOS 15 here in case people want to try but don't want to / couldn't build it https://github.com/lazywei/AeroSpace/actions/runs/14296502264

First impression is snappy fast comparing to previous version. Though previously some slowness only shows up after a while. But at least things like resizing windows seems to be much responsive.

Thanks @nikitabobko so much for working on it.

**Comment by FelixLisczyk** [comment]
Thank you very much for fixing this issue! I can confirm that AeroSpace remains responsive when Xcode pauses my macOS applications at a breakpoint. When the CPU is at 100% during compilation, there is still a delay of 2–4 seconds, but it is much better than before.

**Comment by FelixLisczyk** [comment]
I've noticed some unusual behavior with the new version of AeroSpace when used with Raycast. Initially, everything works as expected. However, after some time, AeroSpace stops automatically switching spaces when I activate an already running app using Raycast. Additionally, the floating Raycast Notes window no longer moves between workspaces. When I open apps through Alfred or Finder, AeroSpace switches workspaces correctly.

After reverting to the previous version of AeroSpace, the problem has not happened again. So far, I have not been able to determine the cause.

**Comment by zhenpewu** [comment]
I just tried the version published by lazywei. It seems my iTerm2 (hotkey floating window) keeps getting assigned to other monitors / workspace in another monitor. After setting it back to the current release version, this issue seems gone.

```
[[on-window-detected]]
if.app-id = 'com.googlecode.iterm2'
run = 'layout floating'
```



<img width="1313" alt="Image" src="https://github.com/user-attachments/assets/e0638d86-20fa-4548-9246-1e141abd2f99" />

**Comment by nikitabobko** [comment]
I've published the 0.18.0 release with the performance fixes. I optimistically close the issue, but I will keep an eye on the feedback

Future optimization directions:
- Investigate Godot https://github.com/nikitabobko/AeroSpace/issues/218
- Avoid accessibility requests to windows on invisible workspaces https://github.com/nikitabobko/AeroSpace/issues/545

**Comment by nikitabobko** [comment]
@FelixLisczyk @zhenpewu I extracted your reports into a separate discussions: https://github.com/nikitabobko/AeroSpace/discussions/1281 https://github.com/nikitabobko/AeroSpace/discussions/1282

**Comment by Morette** [comment]
I'm facing the same isse as someone reported. If I don't change space for like 10s, the next will have a delay of 2/3s but if constantly change, no delay anymore, only if I stop changing again.

I'm not using IntelliJ or any other JetBrains tooling at the moment and my cpu usage is around 10 / 20% and ram 70 / 80%. I'm using a Mac Pro Intel I7 Chip.

I'm using this version
```
aerospace CLI client version: 0.18.2-Beta 7feb63d1ce9aff0d6fdaa13b42888d5b45f5dbb5
AeroSpace.app server version: 0.18.2-Beta 7feb63d1ce9aff0d6fdaa13b42888d5b45f5dbb5
```

This is my config
```
after-login-command = []

after-startup-command = [
  'exec-and-forget sketchybar',
  'exec-and-forget borders active_color=0xffe1e3e4 inactive_color=0xff494d64 width=5.0'
]

exec-on-workspace-change = [
    '/bin/bash', 
    '-c',
    '/usr/local/bin/sketchybar --trigger aerospace_workspace_change FOCUSED_WORKSPACE=$AEROSPACE_FOCUSED_WORKSPACE EMPTY_WORKSPACES=$AEROSPACE_EMPTY_WORKSPACES'
]

start-at-login = true

enable-normalization-flatten-containers = true
enable-normalization-opposite-orientation-for-nested-containers = true

accordion-padding = 30
default-root-container-layout = 'tiles'
default-root-container-orientation = 'auto'

on-focused-monitor-changed = ['move-mouse monitor-lazy-center']
automatically-unhide-macos-hidden-apps = false

[key-mapping]
    preset = 'qwerty'

[gaps]
    inner.horizontal = 15
    inner.vertical =   15
    outer.left =       15
    outer.bottom =     15
    outer.top =        50
    outer.right =      15

[mode.main.binding]
    alt-slash = 'layout tiles horizontal vertical'
    alt-comma = 'layout accordion horizontal vertical'

    cmd-left = 'focus left'
    cmd-down = 'focus down'
    cmd-up = 'focus up'
    cmd-right = 'focus right'

    cmd-ctrl-left = 'move left'
    cmd-ctrl-down = 'move down'
    cmd-ctrl-up = 'move up'
    cmd-ctrl-right = 'move right'

    alt-minus = 'resize smart -50'
    alt-equal = 'resize smart +50'

    alt-1 = 'workspace 1'
    alt-2 = 'workspace 2'
    alt-3 = 'workspace 3'
    alt-4 = 'workspace 4'
    alt-5 = 'workspace 5'
    alt-6 = 'workspace 6'
    alt-7 = 'workspace 7'
    alt-8 = 'workspace 8'
    alt-9 = 'workspace 9'
    alt-c = 'workspace C'
    alt-s = 'workspace S'
    alt-t = 'workspace T'
    alt-w = 'workspace W'
    alt-z = 'workspace Z'

    alt-shift-1 = ['move-node-to-workspace 1', 'workspace 1']
    alt-shift-2 = ['move-node-to-workspace 2', 'workspace 2']
    alt-shift-3 = ['move-node-to-workspace 3', 'workspace 3']
    alt-shift-4 = ['move-node-to-workspace 4', 'workspace 4']
    alt-shift-5 = ['move-node-to-workspace 5', 'workspace 5']
    alt-shift-6 = ['move-node-to-workspace 6', 'workspace 6']
    alt-shift-7 = ['move-node-to-workspace 7', 'workspace 7']
    alt-shift-8 = ['move-node-to-workspace 8', 'workspace 8']
    alt-shift-9 = ['move-node-to-workspace 9', 'workspace 9']
    alt-shift-c = ['move-node-to-workspace C', 'workspace C']
    alt-shift-s = ['move-node-to-workspace S', 'workspace S']
    alt-shift-t = ['move-node-to-workspace T', 'workspace T']
    alt-shift-w = ['move-node-to-workspace W', 'workspace W']
    alt-shift-z = ['move-node-to-workspace Z', 'workspace Z']

    alt-tab = 'workspace-back-and-forth'
    alt-shift-tab = 'move-workspace-to-monitor --wrap-around next'
    alt-shift-semicolon = 'mode service'

[workspace-to-monitor-force-assignment]
    1 = 'main'
    2 = 'secondary'
    3 = 'main'
    4 = 'secondary'
    5 = 'main'
    6 = 'secondary'
    7 = 'main'
    8 = 'secondary'
    9 = 'main'

[mode.service.binding]
    esc = ['reload-config', 'mode main']
    r = ['flatten-workspace-tree', 'mode main']
    f = ['layout floating tiling', 'mode main']
    backspace = ['close-all-windows-but-current', 'mode main']
    alt-shift-h = ['join-with left', 'mode main']
    alt-shift-j = ['join-with down', 'mode main']
    alt-shift-k = ['join-with up', 'mode main']
    alt-shift-l = ['join-with right', 'mode main']
    down = 'volume down'
    up = 'volume up'
    shift-down = ['volume set 0', 'mode main']
```

Something strange is that this only occurs in single monitor, using dual doens't happen. 
tried removing this config, but didn't solve

`[workspace-to-monitor-force-assignment]`

**Comment by shenweijiekdel** [comment]
was able to reproduce this issue on my MacBook Pro 14 M4: when IntelliJ IDEA is initially opened and processing indexing tasks, there is a brief delay in shifting focus away from IDEA. The behavior returns to normal once IDEA has fully loaded.

**Comment by qinabu** [comment]
0.17.1 version from homebrew doesn't have this problem.
```
brew uninstall aerospace
brew install aerospace@0.17.1
```
and everything works fine

**Comment by ifkit** [comment]
0.19.0-Beta still freezes in the same way
> If I don't change space for like 10s, the next will have a delay of 2/3s but if constantly change, no delay anymore, only if I stop changing again.

**Comment by npv12** [comment]
I have noticed the delays and perf regression when I am using my intel mac but I do not have any perf issues when on m3 mac. Both have similar set of applications running and I can confirm the cpu/mem usage isn't high

**Comment by nikitabobko** [comment]
I'm locking this issue to avoid spamming people's inboxes.

If you observe any performance issues, please consider opening a separate discussion. If you want to submit a performance problem, it'd be wise to attach an `Instruments.app` profiler snapshot. I've explained it here on how to do that: https://github.com/nikitabobko/AeroSpace/discussions/1555
